src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/DefaultRazorFormattingService.cs

```diff
@@ -0,0 +1,513 @@
+            var changes = edits.Select(e => e.AsTextChange(originalText)).Where(c => originalCodeBlockSpan.Contains(c.Span)).ToArray();
+            if (changes.Length == 0)
+            {
+                return originalText;
+            }
+
+            // Apply the C# edits to the document.
+            var changedText = originalText.WithChanges(changes);
+            TrackChangeInSpan(originalText, originalCodeBlockSpan, changedText, out var changedCodeBlockSpan, out var changeEncompassingSpan);
+
+            // We now have the changed document with C# edits. But it might be indented more/less than what we want depending on the context.
+            // So, we want to bring each line to the right level of indentation based on where the block is in the document.
+            // We also need to only do this for the lines that are part of the input range to respect range formatting.
+            var desiredIndentationLevel = context.Indentations[(int)codeBlockRange.Start.Line].IndentationLevel + 1;
+            var editsToApply = new List<TextChange>();
+            var inputSpan = context.Range.AsTextSpan(originalText);
+            TrackChangeInSpan(originalText, inputSpan, changedText, out var changedInputSpan, out _);
+            var changedInputRange = changedInputSpan.AsRange(changedText);
+
+            for (var i = (int)changedInputRange.Start.Line; i <= changedInputRange.End.Line; i++)
+            {
+                var line = changedText.Lines[i];
+                if (line.Span.Length == 0)
+                {
+                    // Empty line. C# formatter didn't remove it so we won't either.
+                    continue;
+                }
+
+                if (!changedCodeBlockSpan.Contains(line.Start))
+                {
+                    // Defensive check to make sure we're not handling lines that are not part of the current code block.
+                    continue;
+                }
+
+                var leadingWhitespace = line.GetLeadingWhitespace();
+                var minCSharpIndentLength = context.Options.TabSize * minCSharpIndentLevel;
+                if (leadingWhitespace.Length < minCSharpIndentLength)
+                {
+                    // For whatever reason, the C# formatter decided to not indent this. Leave it as is.
+                    continue;
+                }
+                else
+                {
+                    var effectiveDesiredIndentationLevel = desiredIndentationLevel - minCSharpIndentLevel;
+                    if (effectiveDesiredIndentationLevel < 0)
+                    {
+                        // This means that we need to unindent.
+                        var length = Math.Abs(effectiveDesiredIndentationLevel * context.Options.TabSize);
+                        var span = new TextSpan(line.Start, (int)length);
+                        editsToApply.Add(new TextChange(span, string.Empty));
```

Comment:
Hmm, i'm not understanding how this works. How do you unindent a line to the proper spacing  without knowing the lines spacing? I would have imagined this code would have been taking the current indentation into account.

Reply:
Yes I know its shocking :D. But I am trusting the C# formatter to have done the right thing

Reply:
If the C# formatter didn't indent a certain line to at least the minimum that we expect, it will be taken care of by the `if` part and won't reach here.

Reply:
lolol